### DISCUSSION TITLE
Would it be possible to emulate triggers?

### DISCUSSION THREAD

**Discussion by kiwicopple** [supabase/supabase#26629] [Contribute to Supabase]
Foreign Tables don't support triggers. Would it be possible to emulate them in Wrappers?

See this thread for more context: https://x.com/vimtor_/status/1735244020421455906?s=20

<img width="838" alt="image" src="https://github.com/supabase/wrappers/assets/10214025/1285e4ca-d91e-4bcc-938a-aa6081a117fa">

**Comment by vimtor** [comment]
To give a real-world example, I developed a news site for a client using Supabase and Stripe. On the site, users could buy one-off premium articles to read. The way we achieved this was by listening for product purchases via Stripe webhooks and inserting a new row into `articles_purchases` using the metadata of the Stripe transaction. At the time, you released the first version of Supabase Wrappers, and I thought to myself, "Man, wouldn't it be cool to create a trigger for this?"

**Comment by burmecia** [comment]
Emulating trigger on foreign table would need to constantly monitor source table for changes, that is technically possible but may need time efficiency tradeoff. Actually, if we can build that source monitoring mechanism, it will bring us more possibilities. For example, when new source data detected, we can do 

- insert the new record to another table, either local or foreign table
- send notification to destination
- call API endpoint
- etc.

Basically, we just build a mini CDC (change data capture) and ETL/ELT pipeline around the foreign table. That's the most common usage scenario for FDW, uses normally build it by themselves but this feature will make such task simpler.

**Comment by Rubenburdin** [comment]
Hey vimtor, coming back to this in 2025 since the FDW trigger limitation still exists and I've seen this exact pattern cause headaches for teams.
The workaround most folks end up building is exactly what burmecia described: a custom CDC layer that polls or listens for changes on the source system and pushes them into Supabase. For Stripe specifically, you're stuck with webhooks, which means handling retries, idempotency, and the occasional out of order event yourself. It works, but it's a lot of plumbing to maintain for what should be a straightforward sync.
If you're still running into this kind of problem, [Stacksync](https://www.stacksync.com/) was built precisely for these scenarios. It handles the CDC and bi-directional sync between external systems and Postgres without needing to wire up webhook handlers or poll foreign tables. You get the trigger-like behavior you wanted, just running outside the database instead of inside it.
Curious if you ended up building something custom for the Stripe case or found another path. Hope this helps either way.
